### PR TITLE
BUGFIX - Method setter resolutions in code generator

### PR DESCRIPTION
**Pull request by dskatz**
I tried to keep this PR small, however some fundamental changes need to be made to address type and package resolution.  In order to fix the bugs entirely I opted to keep this in a single changeset.

## Package Index
The first fundamental change is how an index is built of all types required during the code generation process.  Previously `enabledTypes` was built during generation.  This has been replaced entirely by the new `PackageIndex` and `index` package.  Now the index is built prior to generation.  The index is comprised of the mapping of the absolute type name `package/v1.Type` to the `types.Type`.

As a side affect of this change type names need to be cleaned `cleanTypeName()`, since the absolute type name can include `[]` or `*` or both depending on how the type is referenced in the code. e.g. It can be a slice, pointer, both, or none.

## Enum / Alias handling

The `Underlying` type does not correctly resolve.  It walks the tree all the way to the primitive.  e.g. 
```golang
type MyType string
type MyAlias MyType
```

The Underling type of `MyAlias` will not be `MyType`, the parser walks the tree all the way to the primitive `string`.  Therefore I introduce a new `ref` tag on the builder.  This is needed to give a type hint to the code generator when creating setters.

## Package Root Resolution

The package root resolves by discovering the go.mod module path.  This is used for cross module imports.  For example

- pkg/a - Contains auto generated code
- pkg/b - Contains auto generated code

If a type in  `pkg/a` contains a type that is tagged for auto generation by `pkg/b`,  `pkg/b` will be imported into `pkg/a. 

This is most common with Kubernetes CRDS that use K8s base types.

## GenerateType State Machine
- new edge cases were identified to aid code generation. 
- combinations of Struct and Pointers 
- Alias and Builtin
- All case statements now check if a type is enabled.
- Debug logging has been added to help trace the state machine.

### CODE REVIEW

File: pkg/generators/index/index.go [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+			for _, m := range t.Members {
+				if m.Embedded {
+					if m.Name == "StructField" {
+						continue
+					}
+					if _, ok := index[m.Type.String()]; !ok {
+						index[m.Type.String()] = t
+						log.Debugf("Indexing %s -> (%s, %s) -- Package -> %s(%s)", m.Type.String(), m.Name, m.Type.Name, pkg.Path, pkg.SourcePath)
+					}
+				}
```
[the remark this answers is not in the document]
Comment (dskatz) [reply pkg/generators/index/index.go:23]: This was dead logic.  Removing.

File: pkg/generators/snippets/setters.go [hunk 2]
```diff
@@ -175,6 +175,46 @@ func (o $.pointer$$.type|raw$) $.funcName$(in ...*$.inputType|raw$) $.pointer$$.
...
+		o.$.memberAccessor$ = append(o.$.memberAccessor$,  $slice (.enumType|raw) 2$(elem))
+	}
+	return o
+}
+
+`
+	return raw, args
+}
+
+func (s *Setter) GenerateSetterForEmbeddedSlicePointer(member types.Member, inputType *types.Type) (string, generator.Args) {
```
[the remark this answers is not in the document]
Comment (dskatz) [reply pkg/generators/snippets/setters.go:197]: My bad, missed a test for this.

**Review by whwalter — APPROVED** [review]
LGTM but this is a breaking change.

**Comment by dskatz** [comment]
> LGTM but this is a breaking change.

Yes, the version will get bumped to v0.2.0.

**Pull request merged**